clang/include/clang/AST/StmtOpenACC.h

```diff
@@ -101,24 +113,45 @@ class OpenACCAssociatedStmtConstruct : public OpenACCConstructStmt {
 /// those three, as they are semantically identical, and have only minor
 /// differences in the permitted list of clauses, which can be differentiated by
 /// the 'Kind'.
-class OpenACCComputeConstruct : public OpenACCAssociatedStmtConstruct {
+class OpenACCComputeConstruct final
+    : public OpenACCAssociatedStmtConstruct,
+      public llvm::TrailingObjects<OpenACCComputeConstruct,
+                                   const OpenACCClause *> {
   friend class ASTStmtWriter;
   friend class ASTStmtReader;
   friend class ASTContext;
-  OpenACCComputeConstruct()
-      : OpenACCAssociatedStmtConstruct(
-            OpenACCComputeConstructClass, OpenACCDirectiveKind::Invalid,
-            SourceLocation{}, SourceLocation{}, /*AssociatedStmt=*/nullptr) {}
+  OpenACCComputeConstruct(unsigned NumClauses)
+      : OpenACCAssociatedStmtConstruct(OpenACCComputeConstructClass,
+                                       OpenACCDirectiveKind::Invalid,
+                                       SourceLocation{}, SourceLocation{},
+                                       /*AssociatedStmt=*/nullptr) {
+    // We cannot send the TrailingObjects storage to the base class (which holds
+    // a reference to the data) until it is constructed, so we have to set it
+    // separately here.
+    memset(getTrailingObjects<const OpenACCClause *>(), 0,
+           NumClauses * sizeof(const OpenACCClause *));
```

Comment:
I think there is something like uninitialized_init or something like , which can be used instead

Reply:
Do you mean https://en.cppreference.com/w/cpp/memory/uninitialized_default_construct ?  Or is there something else you mean?

`std::uninitialized_default_construct` unfortunately 'default constructs', which means I get indeterminate values instead of nullptr.

Reply:
I meant something like std::unitialized_copy

Reply:
Ah!  That DOES seem like it is useful in the other init here.  Perhaps `std::uninitialized_value_construct` is useful for the memcpy.  Patch incoming :)

Reply:
I just thought that memset can be removed here and instead use std::unitialized_copy in setClauseList. But up to you, this is just a suggestion

Reply:
I think there is value in setting this to '0', but I've used your suggestion in another way.  I don't think 'setClauseList' can do the copy (as it is just initializing the ArrayRef to the pointer+size pair), but I'm also going to use `std::uninitialized_copy` for the setting of the trailing storage.  

The 'serialization' means the clause `MutableArrayRef` needs to be set in the 'create empty' case.  Let me know what you think about the next version of this patch (more suggestions appreciated!).